providers/backend/backend_test.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/openpubkey/openpubkey/util"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/maps"
+)
+
+func TestSimpleBackendOverride(t *testing.T) {
+
+	issuer := "https://accounts.example.com/"
+	mockBackend, err := NewMockProviderBackend(issuer, 3)
+	require.NoError(t, err)
+
+	expSigningKey, expKeyID, expRecord := mockBackend.RandomSigningKey()
+
+	idTokenTemplate := IDTokenTemplate{
+		CommitFunc:           AddAudCommit,
+		Issuer:               issuer,
+		Nonce:                "empty",
+		NoNonce:              false,
+		Aud:                  "also me",
+		KeyID:                expKeyID,
+		NoKeyID:              false,
+		Alg:                  expRecord.Alg,
+		NoAlg:                false,
+		ExtraClaims:          map[string]any{"extraClaim": "extraClaimValue"},
+		ExtraProtectedClaims: map[string]any{"extraHeader": "extraheaderValue"},
+		SigningKey:           expSigningKey,
+	}
+	mockBackend.SetIDTokenTemplate(&idTokenTemplate)
+
+	cicHash := util.Base64EncodeForJWT([]byte("0123456789ABCDEF0123456789ABCDEF"))
+	idt, err := mockBackend.RequestTokenOverrideFunc(string(cicHash))
+	require.NoError(t, err)
+	require.NotNil(t, idt)
+
+	record, err := mockBackend.GetPublicKeyFinder().ByToken(context.Background(), issuer, idt)
+	require.NoError(t, err)
+
+	payload, err := jws.Verify(idt, jws.WithKey(jwa.KeyAlgorithmFrom(record.Alg), record.PublicKey))
+	require.NoError(t, err)
+	require.Contains(t, string(payload), string(cicHash))
+}
+
+func TestKeySetCreatorsConvenience(t *testing.T) {
+	issuer := "https://accounts.example.com/"
+	skSet, recordSet, err := CreateRS256KeySet(issuer, 2)
+	require.NoError(t, err)
+	CheckKeySets(t, "Happy case: CreateRS256KeySet", issuer, "RS256", skSet, recordSet)
+
+	skSet, recordSet, err = CreateES256KeySet(issuer, 2)
+	require.NoError(t, err)
+	CheckKeySets(t, "Happy case: CreateES256KeySet", issuer, "ES256", skSet, recordSet)
+}
+
+func TestKeySetCreators(t *testing.T) {
+	issuerA := "https://accounts.example.com/"
+	issuerB := "https://diff-accounts.example.com/"
+
+	for numKeys := 1; numKeys < 3; numKeys++ {
+		testCases := []struct {
+			name     string
+			issuer   string
+			alg      string
+			expError string
+		}{
+			{name: fmt.Sprintf("Happy case (RS256): %d key(s)", numKeys), issuer: issuerA, alg: "RS256",
+				expError: "",
+			},
+			{name: fmt.Sprintf("Happy case (ES256): %d key(s)", numKeys), issuer: issuerB, alg: "ES256",
+				expError: "",
+			},
+			{name: fmt.Sprintf("Unsupported alg (ZZ404): %d key(s)", numKeys), issuer: issuerB, alg: "ZZ404",
+				expError: "unsupported alg",
+			},
+		}
+		for _, tc := range testCases {
+			// skSet - set of signingKeys, recordSet - set of publicKeyRecords
+			skSet, recordSet, err := CreateKeySet(tc.issuer, tc.alg, numKeys)
+			if tc.expError != "" {
+				require.ErrorContains(t, err, tc.expError, tc.name)
+			} else {
+				require.NoError(t, err, tc.name)
+			}
+			CheckKeySets(t, tc.name, tc.issuer, tc.alg, skSet, recordSet)
+		}
```

Comment:
Sometimes its helpful to call the test cases as subtests `tc.Run()` for error handling in the test results

```suggestion
	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			// skSet - set of signingKeys, recordSet - set of publicKeyRecords
			skSet, recordSet, err := CreateKeySet(tc.issuer, tc.alg, numKeys)
			if tc.expError != "" {
				require.ErrorContains(t, err, tc.expError, tc.name)
			} else {
				require.NoError(t, err, tc.name)
			}
			CheckKeySets(t, tc.name, tc.issuer, tc.alg, skSet, recordSet)
		})
	}
```